packages/core/src/application.ts

```diff
@@ -91,19 +98,59 @@ export class Application extends Context {
 
     this._httpHandler = new HttpHandler(this);
     for (const b of this.find('controllers.*')) {
+      const controllerName = b.key.replace(/^controllers\./, '');
       const ctor = b.valueConstructor;
       if (!ctor) {
-        throw new Error(`The controller ${b.key} was not bound via .toClass()`);
+        throw new Error(
+          `The controller ${controllerName} was not bound via .toClass()`);
       }
       const apiSpec = getApiSpec(ctor);
-      this._httpHandler.registerController(b.key, apiSpec);
+      if (!apiSpec) {
+        // controller methods are specified through app.api() spec
+        continue;
+      }
+      this._httpHandler.registerController(controllerName, apiSpec);
     }
 
     for (const b of this.find('routes.*')) {
       // TODO(bajtos) should we support routes defined asynchronously?
       const route = this.getSync(b.key);
       this._httpHandler.registerRoute(route);
     }
+
+    // TODO(bajtos) should we support API spec defined asynchronously?
+    const spec: OpenApiSpec = this.getSync('api-spec');
+    for (const path in spec.paths) {
+      for (const verb in spec.paths[path]) {
+        const routeSpec: OperationObject = spec.paths[path][verb];
+        this._setupOperation(verb, path, routeSpec);
+      }
+    }
+  }
+
+  private _setupOperation(verb: string, path: string, spec: OperationObject) {
+    const handler = spec['x-operation'];
```

Comment:
I don't mind either way. Can you discuss with @ritch  please?

Let's leave such change out of scope of this pull request please, `x-operation` was already implemented by #416.